eval/lib/storybook-dev-server.ts

```diff
@@ -0,0 +1,112 @@
+import { createServer } from 'node:net';
+import { x, type Result } from 'tinyexec';
+
+/**
+ * State for a running Storybook dev server instance.
+ */
+export type StorybookDevServer = {
+	port: number;
+	process: Result;
+};
+
+let storybookDevServer: StorybookDevServer | undefined;
+
+/**
+ * Get a random available port by binding to port 0 and then releasing it.
+ * Binding to port 0 tells the OS to assign any available port.
+ * @see https://nodejs.org/api/net.html#serverlistenport-host-backlog-callback
+ */
+export async function getRandomAvailablePort(): Promise<number> {
+	return new Promise<number>((resolve, reject) => {
+		const server = createServer();
+		server.listen(0, () => {
+			const address = server.address();
+			if (address && typeof address === 'object') {
+				const port = address.port;
+				server.close(() => resolve(port));
+			} else {
+				server.close(() => reject(new Error('Failed to get port')));
+			}
+		});
+		server.on('error', reject);
+	});
+}
+
+/**
+ * Wait for Storybook to be ready by listening for the "Storybook ready!" message.
+ */
+export async function waitForStorybook(
+	process: Result,
+	timeout: number = 10_000,
+): Promise<void> {
+	return await new Promise<void>((resolve, reject) => {
```

Comment:
Why are you returning a `new Promise` inside of an async function, instead of using await inside and throwing errors directly? Is it because of the `onExit` callback function, which is passed to the `off,on` event listeners?